raiden/network/transport/matrix/transport.py

```diff
@@ -970,79 +952,68 @@ def _get_room_for_address(self, address: Address, allow_missing_peers=False) ->
-
         assert self._raiden_service is not None
-        address_pair = sorted(
-            [to_normalized_address(address) for address in [address, self._raiden_service.address]]
-        )
-        room_name = make_room_alias(self.network_id, *address_pair)
 
-        # no room with expected name => create one and invite peer
+        # filter peer_candidates
         peer_candidates = [
             self._get_user(user) for user in self._client.search_user_directory(address_hex)
         ]
-
-        # filter peer_candidates
         peers = [user for user in peer_candidates if validate_userid_signature(user) == address]
+
         if not peers and not allow_missing_peers:
             self.log.error("No valid peer found", peer_address=to_checksum_address(address))
             return None
 
+        # filter_private is done in _get_room_ids_for_address
+        room_ids = self._get_room_ids_for_address(address)
+        online_user_ids = self._address_mgr.get_online_user_ids_for_address(address)
+        for room_id in room_ids:
+            room = self._client.rooms[room_id]
+
+            # Ignore global rooms
+            if not self._is_room_global(room):
+                continue
+
+            member_ids = {member.user_id for member in room.get_joined_members()}
+
+            # Handle the case where an empty room was found
+            if not member_ids:
+                for peer in peers:
+                    self._invite_user_to_room_with_retries(peer, room)
+                member_ids = {member.user_id for member in room.get_joined_members()}
+
+            # Do only return a room where no currently online user is listening
+            # before all rooms have been checked for an online peer
+            if member_ids & online_user_ids:
+                self.log.debug(
+                    "Existing room",
+                    room=room,
+                    members=member_ids,
+                    listening_peers=member_ids & online_user_ids,
+                )
+                return room
+            else:
+                self.log.error(
+                    "No online user listening to current room",
+                    room=room,
+                    users_currently_online=online_user_ids,
+                    room_members=member_ids,
+                    peer_address=to_checksum_address(address),
+                )
+                return room
+
+        # no room with expected name => create one and invite peer
         if self._private_rooms:
             room = self._get_private_room(invitees=peers)
         else:
-            room = self._get_public_room(room_name, invitees=peers)
+            room = self._get_public_room(invitees=peers)
 
         peer_ids = self._address_mgr.get_userids_for_address(address)
-        member_ids = {member.user_id for member in room.get_joined_members(force_resync=True)}
+        member_ids = {member.user_id for member in room.get_joined_members()}
```

Comment:
I think the problem here is that we have a lot of overlapping logic, so there's room creation, than there's invitations coupled with room creation (at least in get_public_room) and than we have these "wrapping" invite retries. 
This is all there to prevent us from sending messages to a room with offline or no users, which would lead to expiring transfers (I think?).

So we're trading state safety for speed here as the membership check is sure expensive. 
With #4089 we can recover user_ids and room memberships, so it should be a lot faster. But I agree that decoupling room creation, invites, online checks etc. from each other is also something we should do.

Reply:
With the last commit I removed the invites and retries from `_get_public_room` to separate room creation and invites. Also, I spend some more thought on your comment, and I think the call here makes sence, as  _invite_user_with_retries sleeps in case the user does not join, so the call should return a proper result.